app/components/file/browse/file-explorer/file-explorer-workspace.ts

```diff
@@ -0,0 +1,161 @@
+                    treeNode: node,
+                };
+            });
+        }).shareReplay(1);
+
+        this._openInitalFiles();
+    }
+
+    public navigateTo(path: string, source?: FileSource) {
+        source = this._getSource(source);
+        if (this._currentPath.value === path && this._currentSource.value === source) { return; }
+
+        this._currentPath.next(path);
+        this._currentSource.next(source);
+        source.navigator.getNode(path).subscribe((node) => {
+            if (!node) { return; }
+            if (node.isDirectory) {
+                source.navigator.loadPath(path);
+            } else {
+                this.openFile(node.path, source);
+            }
+        });
+    }
+
+    /**
+     * Go back up one level
+     */
+    public goBack() {
+        const path = this._currentPath.value;
+        if (path === "") { return; }
+        this.navigateTo(path.split("/").slice(0, -1).join("/"), this._currentSource.value);
+    }
+
+    public isFileOpen(path: string, source?: FileSource): boolean {
+        source = this._getSource(source);
+        return Boolean(this._openedFiles.value.find(x => x.path === path && x.source === source));
+    }
+
+    public openFile(path: string, source?: FileSource) {
+        source = this._getSource(source);
+        const openedFiles = this._openedFiles.value;
+        if (!this.isFileOpen(path, source)) {
+            openedFiles.push({
+                source: source,
+                path,
+                loader: source.navigator.getFile(path),
+            });
+        }
+        this._openedFiles.next(openedFiles);
+    }
+
+    /**
+     * Triggered when a tab select to open a file
+     * @param filename File to open
+     *
+     * If filename is null or undefined it will show the file table viewer at the last position
+     */
+    public openFiles(paths: string[], source?: FileSource) {
+        source = this._getSource(source);
+
+        const openedFiles = this._openedFiles.value;
+        for (let path of paths) {
+            if (!this.isFileOpen(path, source)) {
+                openedFiles.push({
+                    source: source,
+                    path,
+                    loader: source.navigator.getFile(path),
+                });
+            }
+        }
+        this._openedFiles.next(openedFiles);
+    }
+
+    public closeFile(path: string, source?: FileSource) {
+        source = this._getSource(source);
+
+        const newOpenedFiles = this._openedFiles.value.filter(x => x.path !== path && x.source === source);
+        this._openedFiles.next(newOpenedFiles);
+    }
+
+    public dispose() {
+        this.sources.forEach(x => x.navigator.dispose());
+    }
+
+    private _getSource(source?: FileSource) {
+        if (source) { return source; }
+        if (this.sources.length === 1) {
+            return this.sources.first();
+        } else {
+            log.error("You must specify a source(FileNavigator) when using multi-source");
```

Comment:
bit of a nitpick but this could go into a const so we only need to change it in one place if needs be.

Reply:
yeah speaking of that I think we should look at internationalization that would make the code cleaner for long messages at least